### PR TITLE
Implement data-sharding in PackedDataset

### PR DESCRIPTION
**Pull request by awaelchli**
Follow-up to #217

The iterable dataset implementation for redpajama so far supported only sharding the data across the dataloader workers, but not globally across all processes in a distributed setting, e.g., Fabric(devices > 1).

### CODE REVIEW

File: train_redpajama.py [hunk 1]
```diff
@@ -226,14 +227,16 @@ def create_dataloader(
...
     shuffle: bool = True,
     seed: int = 12345,
 ) -> DataLoader:
     datasets = []
     for prefix, _ in data_config:
         filenames = glob.glob(os.path.join(data_dir, prefix + "*"))
         dataset = PackedDataset(
-            filenames, n_chunks=10, block_size=block_size, shuffle=shuffle, seed=seed
+            filenames, n_chunks=4, block_size=block_size, shuffle=shuffle, seed=seed,
+            num_processes=fabric.world_size, process_rank=fabric.global_rank,
```
Comment (awaelchli) [inline train_redpajama.py:239]: the n_chunks value here must be smaller than the number of files ending up in each shard. It's unclear to me what exactly the chunks refer to

Comment (lantiga) [reply train_redpajama.py:239]: Say you have 12 files (12 chunks, each chunk containing N blocks of tokens) and you set `n_chunks=4`.
The dataset will pre-load 4 chunks in memory and iterate over the blocks contained in those chunks, optionally shuffling the blocks. If the dataset doesn't find enough chunks to load, it will raise a StopIteration.

I think we should actually give the opportunity to rewind as well for overfit tests etc.

**Review by lantiga — APPROVED** [review]
Looks great
